src/freenet/support/io/InetAddressIpv6FirstComparator.java

```diff
@@ -1,29 +1,71 @@
 package freenet.support.io;
 
+import java.io.IOException;
 import java.net.InetAddress;
 import java.util.Comparator;
 
 import freenet.support.Fields;
+import freenet.support.LRUCache;
 
 /** Comparator for IP addresses that sorts IPv6 before IPv4 to enable
  * selecting the first.
  * @author toad */
 public class InetAddressIpv6FirstComparator implements Comparator<InetAddress> {
 
-	public final static InetAddressIpv6FirstComparator COMPARATOR =
+	// need a cache for reachability to avoid doing NlogN issReachable checks in worst case.
+	public LRUCache<Integer, Boolean> reachabilityCache = new LRUCache<>(1000, 10000);
+
+	public static final InetAddressIpv6FirstComparator COMPARATOR =
 		new InetAddressIpv6FirstComparator();
 
 	@Override
 	public int compare(InetAddress arg0, InetAddress arg1) {
 		if(arg0 == arg1) return 0;
+		// prefer everything over loopback
+		if (!arg0.isLoopbackAddress() && arg1.isLoopbackAddress()) {
+			return -1;
+		} else if (arg0.isLoopbackAddress() && !arg1.isLoopbackAddress()) {
+			return 1;
+		}
+        // prefer LAN routable addresses over link-local addresses
+		if (!arg0.isLinkLocalAddress() && arg1.isLinkLocalAddress()) {
+			return -1;
+		} else if (arg0.isLinkLocalAddress() && !arg1.isLinkLocalAddress()) {
+			return 1;
+		}
 		byte[] bytes0 = arg0.getAddress();
 		byte[] bytes1 = arg1.getAddress();
+        // prefer IPv6 over IPv4
 		if(bytes0.length > bytes1.length)
-			return -1; // IPv6 < IPv4. => prefer IPv6 over IPv4
+			return -1;
 		if(bytes1.length > bytes0.length)
-			return 1; // IPv4 < IPv6.
+			return 1;
+		// prefer reachable over unreachable. TODO: This actually pings all advertised ip addresses. Is that OK?
```

Comment:
prefer reachable over unreachable is OK.  What do you mean by pings all advertised IP addresses? ping as ping (ICMP) or freenet / fred specific UDP connection attempt?

Reply:
PING as ICMP.